src/igraph/io/objects.py

```diff
@@ -0,0 +1,813 @@
+                    edge_attrs[k] = [(idx, v)]
+            m += 1
+        for k, v in edge_attrs.items():
+            edge_attrs[k] = create_list_from_indices(v, m)
+
+        # It may have happened that some vertices were added during
+        # the process
+        if len(vertex_name_map) > n:
+            diff = len(vertex_name_map) - n
+            more = [None] * diff
+            for k, v in vertex_attrs.items():
+                v.extend(more)
+            vertex_attrs[vertex_name_attr] = list(vertex_name_map.values())
+            n = len(vertex_name_map)
+
+        # Create the graph
+        return cls(n, edge_list, directed, {}, vertex_attrs, edge_attrs)
+
+
+def _construct_graph_from_tuple_list(
+    cls,
+    edges,
+    directed=False,
+    vertex_name_attr="name",
+    edge_attrs=None,
+    weights=False,
+):
+    """Constructs a graph from a list-of-tuples representation.
+
+    This representation assumes that the edges of the graph are encoded
+    in a list of tuples (or lists). Each item in the list must have at least
+    two elements, which specify the source and the target vertices of the edge.
+    The remaining elements (if any) specify the edge attributes of that edge,
+    where the names of the edge attributes originate from the C{edge_attrs}
+    list. The names of the vertices will be stored in the vertex attribute
+    given by C{vertex_name_attr}.
+
+    The default parameters of this function are suitable for creating
+    unweighted graphs from lists where each item contains the source vertex
+    and the target vertex. If you have a weighted graph, you can use items
+    where the third item contains the weight of the edge by setting
+    C{edge_attrs} to C{"weight"} or C{["weight"]}. If you have even more
+    edge attributes, add them to the end of each item in the C{edges}
+    list and also specify the corresponding edge attribute names in
+    C{edge_attrs} as a list.
+
+    @param edges: the data source for the edges. This must be a list
+      where each item is a tuple (or list) containing at least two
+      items: the name of the source and the target vertex. Note that
+      names will be assigned to the C{name} vertex attribute (or another
+      vertex attribute if C{vertex_name_attr} is specified), even if
+      all the vertex names in the list are in fact numbers.
+    @param directed: whether the constructed graph will be directed
+    @param vertex_name_attr: the name of the vertex attribute that will
+      contain the vertex names.
+    @param edge_attrs: the names of the edge attributes that are filled
+      with the extra items in the edge list (starting from index 2, since
+      the first two items are the source and target vertices). C{None}
+      means that only the source and target vertices will be extracted
+      from each item. If you pass a string here, it will be wrapped in
+      a list for convenience.
+    @param weights: alternative way to specify that the graph is
```

Comment:
@ntamas I'll deflect this one to you, as I'm not sure what the original intention here was.

I *suspect* it could be a common situation to have two separate data structures for edge source-targets and for weights, and this argument lifts the need to `zip` them together, which is slow in Python. If that's the case, I think the original rationale was fine but now that we have pandas DataFrame support and a bunch of other input constructors we could get rid of this (in a separate PR, since it requires API deprecation).

If you think that's alright we can open an issue about deprecating this and I'll make a separate PR.

Reply:
`zip` is probably not that slow that it would matter; anyhow, the idea was indeed to make it easy to deal with the common scenario of having an edge list + weights only. I'll leave it as is for the time being; we can go through a deprecation cycle if it really bothers someone but I'm okay with keeping the old API.